### PR TITLE
Interval assertions for input data

### LINKED ISSUE #35: Add function to check contiguous age groups for missings

### LINKED ISSUE #46: FEATURE: Split extra flexibility/checks from main aggregation and scaling functions
**Is your feature request related to a problem? Please describe.**
The current `agg` and `scale` functions are too slow when used on large datasets because some of the checks are not efficient.

**Describe the solution you'd like**
Split the extra flexibility and checks into separate functions that can be used as needed.

**Describe alternatives you've considered**
Adding an option to the current functions to use some of the longer checks/flexibility.

### LINKED ISSUE #63: FEATURE: Identify formatting issues before running agg
**Is your feature request related to a problem? Please describe.**
When aggregating data, it is necessary to run agg in a loop using a unique identifier like nid-location-year-sex. When there is an error, the loop stops.

**Describe the solution you'd like**
 It would be nice to have a helper function which could identify these issues with the data beforehand.

### PR DESCRIPTION
**Pull request by chacalle**
## Describe changes

Makes existing `identify_missing_intervals` and `identify_overlapping_intervals` non-exported helper functions.

Creates new `identify_missing_intervals_dt` and `identify_overlapping_intervals_dt` that take input data defined by `id_cols` which matches more realistic input data.

## What issues are related

Fixes #63 
Fixes #46 
Fixes #35

### Packages Repositories

* [X] Have you read the [contributing guidelines](https://github.com/ihmeuw-demographics/packageTemplate/wiki#guide-to-r-package-development) for `ihmeuw-demographics` R packages?
* [X] Have you successfully run `devtools::check()` locally?
* [X] Have you updated or added function (and vignette if applicable) documentation? Did you update the 'man' and 'NAMESPACE' files with `devtools::document()`?
* [X] Have you added in tests for the changes included in the PR?
* [X] Do the changes follow the `ihmeuw-demographics` [code style](https://github.com/ihmeuw-demographics/packageTemplate/wiki/Code-style-guide)?
* [ ] Do the changes need to be immediately included in a new build of [`docker-base`](https://github.com/ihmeuw-demographics/docker-base) or [`docker-internal`](https://github.com/ihmeuw-demographics/docker-internal)? If so follow directions in those repositories to rebuild and redeploy the images.
* [ ] Do the changes require updates to other repositories which use this package? If yes, make the necessary updates in those repos, and consider integration tests for those repositories.
* [ ] If this is a private package did you use Jenkins to rebuild the internal pkgdown site?

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
+export(assert_no_overlapping_intervals_dt)
 export(collapse_common_intervals)
 export(create_agg_tree)
 export(create_scale_tree)
 export(gen_end)
 export(gen_length)
 export(gen_name)
-export(identify_missing_intervals)
-export(identify_overlapping_intervals)
+export(identify_missing_intervals_dt)
```
Comment (erinamay) [inline NAMESPACE:12]: Does the function need to change names?

**Review by erinamay — DISMISSED** [review]

File: R/interval_assertions.R [hunk 2]
```diff
@@ -1,4 +1,182 @@
...
+#'
+#' @return  `identify_missing_intervals_dt` returns a \[`data.table()`\] with
+#'   `id_cols` that are missing expected intervals. If no intervals are missing
+#'   then a zero-row \[`data.table()`\] is returned.
+#'   `assert_no_missing_intervals_dt` returns nothing but throws an error if
+#'   `identify_missing_intervals` returns a non-empty data.table.
+#'
+#' @details
+#' `identify_missing_intervals_dt` works by first identifying each unique set of
+#' intervals in `dt`. Then checks one at at a time the groups of rows of `dt`
```
Comment (hcomfo95) [inline R/interval_assertions.R:31]: remove repeat `at`

File: R/interval_assertions.R [hunk 3]
```diff
@@ -86,7 +262,167 @@ identify_missing_intervals <- function(ints_dt, expected_ints_dt) {
...
+#'
+#' @return  `identify_overlapping_intervals_dt` returns a \[`data.table()`\] with
+#'   `id_cols` that have overlapping intervals. If no intervals are overlapping
+#'   then a zero-row \[`data.table()`\] is returned.
+#'   `assert_no_overlapping_intervals_dt` returns nothing but throws an error if
+#'   `identify_overlapping_intervals_dt` returns a non-empty data.table.
+#'
+#' @details
+#' `identify_overlapping_intervals_dt` works by first identifying each unique
+#' set of intervals in `dt`. Then checks one at at a time the groups of rows
```
Comment (hcomfo95) [inline R/interval_assertions.R:284]: remove repeat `at`

File: R/interval_assertions.R [hunk 4]
```diff
@@ -86,7 +262,167 @@ identify_missing_intervals <- function(ints_dt, expected_ints_dt) {
...
+  })
+  overlapping_intervals_dt <- rbindlist(overlapping_intervals_dt, use.names = TRUE)
+
+  data.table::setcolorder(overlapping_intervals_dt, original_col_order)
+  data.table::setkeyv(overlapping_intervals_dt, original_keys)
+  return(overlapping_intervals_dt)
+}
+
+
+#' @title Helper function to check ift he interval variable has any overlapping
```
Comment (hcomfo95) [inline R/interval_assertions.R:424]: typo

**Review by hcomfo95 — DISMISSED** [review]

File: NAMESPACE (hunk 1 above)
Comment (chacalle) [reply NAMESPACE:12]: Its not changing names. `identify_missing_intervals` is the same, just no longer exported from the package. `identify_missing_intervals_dt` is a new function that will be exported

**Pull request merged**